pallets/process-validation/src/lib.rs

```diff
@@ -79,53 +90,110 @@ pub mod pallet {
 
     /// Storage map definition
     #[pallet::storage]
-    #[pallet::getter(fn processes_by_id_and_version)]
-    pub(super) type ProcessesByIdAndVersion<T: Config> = StorageDoubleMap<
-        _,
-        Blake2_128Concat,
-        T::ProcessIdentifier,
-        Blake2_128Concat,
-        T::ProcessVersion,
-        Process,
-        ValueQuery,
-    >;
+    #[pallet::getter(fn process_model)] // not sure about name, store?, map?
+    pub(super) type ProcessModel<T: Config> =
+        StorageDoubleMap<_, Blake2_128Concat, T::ProcessIdentifier, Blake2_128Concat, i32, Process, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn version_model)]
+    pub(super) type VersionModel<T: Config> = StorageMap<_, Blake2_128Concat, T::ProcessIdentifier, i32, ValueQuery>;
 
     #[pallet::event]
     #[pallet::generate_deposit(pub(super) fn deposit_event)]
     pub enum Event<T: Config> {
-        // TODO: implement correct events for extrinsics including params
-        ProcessCreated,
-        ProcessDisabled,
+        // id, version, restrictions, is_new
+        ProcessCreated(T::ProcessIdentifier, i32, Vec<Restriction>, bool),
+        //id, version, updated
+        ProcessDisabled(T::ProcessIdentifier, i32, bool),
     }
 
     #[pallet::error]
     pub enum Error<T> {
-        // TODO: implement errors for extrinsics
+        NonExistingProcess,
+        AlreadyDisabled,
+        InvalidVersion,
     }
 
     // The pallet's dispatchable functions.
     #[pallet::call]
     impl<T: Config> Pallet<T> {
-        // TODO: implement create_process with correct parameters and impl
         #[pallet::weight(T::WeightInfo::create_process())]
-        pub(super) fn create_process(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            // Check it was signed and get the signer
+        pub(super) fn create_process(
+            origin: OriginFor<T>,
+            id: T::ProcessIdentifier,
+            restrictions: Vec<Restriction>,
+        ) -> DispatchResultWithPostInfo {
             T::CreateProcessOrigin::ensure_origin(origin)?;
+            let new_version: i32 = Pallet::<T>::_update_version(id.clone()).unwrap();
+            Pallet::<T>::_persist_process(&id, &new_version, restrictions.clone());
 
-            Self::deposit_event(Event::ProcessCreated);
-            Ok(().into())
+            // wrap event data into a struct?
```

Comment:
Could do but probably the most idiomatic way of assigning meaning to the values in an event is with event metadata as used in the symmetric key pallet:

```rust
#[pallet::event]
#[pallet::metadata(Vec<u8> = "Key")]
#[pallet::generate_deposit(pub(super) fn deposit_event)]
pub enum Event<T: Config> {
    // key was updated.
    UpdateKey(Vec<u8>),
}
```

Reply:
This is what I wanted!